libs/optimizer/src/lib/optimizer.ts

```diff
@@ -0,0 +1,348 @@
+}
+
+export enum CircuitType {
+  OpenDistance = 'Open distance',
+  FlatTriangle = 'Flat triangle',
+  FaiTriangle = 'Fai triangle',
+  OutAndReturn = 'Out and return',
+}
+
+export interface OptimizationResult {
+  /**
+   * the score for the track in the given league
+   */
+  score: number;
+  /**
+   * the length of the optimized track in kms
+   */
+  lengthKm: number;
+  /**
+   * multiplier for computing score. score = lengthKm * multiplier
+   */
+  multiplier: number;
+  /**
+   * type of the optimized track
+   */
+  circuit?: CircuitType;
+  /**
+   * if applicable, distance in m for closing the circuit
+   */
+  closingRadius?: number;
+  /**
+   * indices of solutions points in
+   */
+  solutionIndices: number[];
+  /**
+   * the result is optimal (no need to get a next result of Iterator<OptimizationResult, OptimizationResult>)
+   */
+  optimal: boolean;
+}
+
+const ZERO_SCORE: OptimizationResult = {
+  score: 0,
+  lengthKm: 0,
+  multiplier: 0,
+  solutionIndices: [],
+  optimal: true,
+};
+
+/**
+ * returns an iterative optimizer that computes iteratively the score for the flight. At each iteration, the score
+ * should be a better solutions.
+ * @param request the OptimizationRequest. if request.options is undefined, then there will be one iteration, and the result
+ *                will be the best solution
+ * @param scoringRules the ScoringRules to apply for computation
+ * @return an Iterator over the successive OptimizationResult
+ * @see README.md
+ */
+export function* getOptimizer(
+  request: OptimizationRequest,
+  scoringRules: ScoringRules,
+): Iterator<OptimizationResult, OptimizationResult> {
+  if (request.track.points.length == 0) {
+    console.warn('Empty track received in optimization request. Returns a 0 score');
+    return ZERO_SCORE;
+  }
+  const originalTrack = request.track;
+  const solverTrack = buildValidTrackForSolver(originalTrack);
+  const flight = toIgcFile(solverTrack);
+  const solverScoringRules = toScoringRules(scoringRules);
+  const options = toOptions(request.options);
+  const solutionIterator = solver(flight, solverScoringRules || {}, options);
+  while (true) {
+    const solution = solutionIterator.next();
+    if (solution.done) {
+      console.debug('solution', JSON.stringify(solution.value, undefined, 2));
+      return toResult(solution.value, originalTrack, solverTrack);
+    }
+    yield toResult(solution.value, originalTrack, solverTrack);
+  }
+}
+
+/**
+ * the solver requires at least 5 points, so if there is not enough points,
+ * we create points between existing ones
+ */
+function buildValidTrackForSolver(track: ScoringTrack) {
+  if (track.points.length >= MIN_POINTS) {
+    return track;
+  }
+  console.debug(`not enough points (${track.points.length}) in track. Interpolate intermediate points`);
+  // avoid potential future bugs with a defensive copy to avoid unwanted mutations on inputs.
+  let newTrack: ScoringTrack = deepCopy(track);
```

Comment:
nit:
- there is no need for an explicit type here
- you can write `track = deepCopy(track);
- the comment does not really add value.

Reply:
Ok. Changed